AccessHandle.md

```diff
@@ -0,0 +1,206 @@
+
+## Introduction
+
+We propose augmenting the Origin Private File System (OPFS) with a new surface
+that brings very performant access to data. This new surface differs from
+existing ones by offering in-place and exclusive write access to a file’s
+content. This change, along with the ability to consistently read unflushed
+modifications and the availability of a synchronous variant on dedicated
+workers, significantly improves performance and unblocks new use cases for the
+File System Access API.
+
+More concretely, we would add a *createAccessHandle()* method to the
+*FileSystemFileHandle* object. It would return an *AccessHandle* that contains
+a [duplex stream](https://streams.spec.whatwg.org/#other-specs-duplex) and
+auxiliary methods. The readable/writable pair in the duplex stream communicates
+with the same backing file, allowing the user to read unflushed contents.
+Another new method, *createSyncAccessHandle()*, would only be exposed on Worker
+threads. This method would offer a more buffer-based surface for reading and
+writing. The creation of AccessHandle also creates a lock that prevents write
+access to the file across (and within the same) execution contexts.
+
+This proposal is part of our effort to merge [Storage Foundation
+API](https://github.com/WICG/storage-foundation-api-explainer) and File System
+Access API. For more context the origins of this proposal, and alternatives
+considered, please check out: [Merging Storage Foundation API and the Origin
+Private File
+System](https://docs.google.com/document/d/121OZpRk7bKSF7qU3kQLqAEUVSNxqREnE98malHYwWec),
+[Recommendation for Augmented
+OPFS](https://docs.google.com/document/d/1g7ZCqZ5NdiU7oqyCpsc2iZ7rRAY1ZXO-9VoG4LfP7fM).
+
+## Goals & Use Cases
+
+Our goal is to give developers flexibility by providing generic, simple, and
+performant primitives upon which they can build higher-level storage
+components. The new surface is particularly well suited for Wasm-based
+libraries and applications that want to use custom storage algorithms to
+fine-tune execution speed and memory usage.
+
+A few examples of what could be done with *AccessHandles*:
+
+*   Distribute a performant Wasm port of SQLite. This gives developers the
+    ability to use a persistent and fast SQL engine without having to rely on
+    the deprecated WebSQL API.
+*   Allow a music production website to operate on large amounts of media, by
+    relying on the new surface's performance and direct buffered access to
+    offload sound segments to disk instead of holding them in memory.
+*   Provide a fast and persistent [Emscripten](https://emscripten.org/)
+    filesystem to act as generic and easily accessible storage for Wasm.
+
+## Non-goals
+
+This proposal is focused only on additions to the [Origin Private File
+System](https://wicg.github.io/file-system-access/#sandboxed-filesystem), and
+doesn't currently consider changes to the rest of File System Access API or how
```

Comment:
While I know you're focusing on the origin private use cases specifically for this proposal, I think it would be worthwhile to explore what parts of this API could be exposed on arbitrary file handles as well. I believe there are a number of use cases that could be addressed by adding a couple of options to the createAccessHandle methods to slightly tweak its behavior.

Straw-man proposal `createAccessHandle({inPlace: true, locking: true})` would match what you've described here, and be only available for origin private files. 

On the other hand inPlace: false could then work with the same API surface for arbitrary file handles. While that would of course duplicate some API surface from the existing createWritable method, being able to read what has been written before it is flushed/comitted seems useful, and if this new API will be built on top of some better seekable streams primitive that might be a benefit as well.

Not sure if a non-locking version for the origin private file system makes sense, or is desirable, but for arbitrary files I think it would be useful to be able to create both locking and non-locking access handles.

While all that can probably be added after the fact, we can't really change the defaults for such options easily, so while it might perhaps make sense to initially only support a "inPlace: true, locking: true" mode (and throw/reject for any other options), I don't think that should be the default flags if we actually want to pursue a more unified API.

Reply:
That's a good point, I agree.

I've added a "mode" parameter to createAccessHandle, and an open question describing why it's there and what other values it could have. Please take a look and let me know if this covers what you meant!